hoomd/output/csv_backend.py

```diff
@@ -0,0 +1,190 @@
+                                   self.max_decimals_pretty)
+                else:
+                    decimals = max(self.precision - digit_guess + 1, 0)
+                type_fmt = "." + str(decimals) + "f"
+                return self._num_format.format(value,
+                                               width=column_width,
+                                               type=type_fmt)
+
+    def format_str(self, value, column_width):
+        if self.pretty and len(value) > column_width:
+            truncate_to = max(1, column_width - 2)
+            return self._str_format.format(value[-truncate_to:],
+                                           width=column_width)
+        else:
+            return self._str_format.format(value, width=column_width)
+
+
+class _CSVInternal(_InternalCustomAction):
+    """Implements the logic for a simple text based logger backend."""
+
+    def __init__(self, logger, output=stdout, sep='.',
+                 pretty=True, max_precision=10, max_len_namespace=None):
+        # Only accept loggers with scalar and string quantities
+        flags = set(logger.flags)
+        if flags.difference(['scalar', 'string']) != set() \
+                or 'scalar' not in flags:
+            raise ValueError("Given Logger must have the scalar flag set.")
+        else:
+            self._logger = logger
+            self._sep = sep
+            # Ensure that columns are always at least ten characters
+            self._min_width = max(10, 10 if pretty else max_precision + 6)
+            # Used to truncate namespaces
+            self._max_len_namespace = max_len_namespace
+            # Records the current keys and their lengths
+            self._cur_headers = dict()
+            self._fmt = _Formatter(pretty, max_precision)
+            # the output file or stdout
+            self._output = output
+
+    def attach(self, simulation):
+        pass
+
+    def detach(self):
+        pass
+
+    def log(self):
+        # Get logger output in {"module{self._sep}class": value} form
+        if self._max_len_namespace is None:
+            output_dict = {
+                self._sep.join(key): value[0]
+                for key, value in dict_flatten(self._logger.log()).items()
+                if value[1] in {'string', 'scalar'}
+            }
+        else:
+            output_dict = {
+                self._sep.join(key[-self._max_len_namespace:]): value[0]
+                for key, value in dict_flatten(self._logger.log()).items()
+                if value[1] in {'string', 'scalar'}
+            }
+        new_keys = output_dict.keys()
+        if new_keys != self._cur_headers.keys():
+            self._cur_headers = {key: max(len(key), self._min_width)
+                                 for key in new_keys}
+            self._write_header()
+
+        self._write_row(output_dict)
+
+    def _write_header(self):
+        headers = self._cur_headers
+        self._output.write(
```

Comment:
Write output only on rank 0.

Also, when using actual file backed output, we will need to provide some mechanism or example showing how to open files only on the root rank.

Reply:
Good observation, and yes we will. I imagine that this will have to be something we just show users. We could also duck type the output to accept a string as well which we can open internally with an optional mode argument.